modules/mssql/connection.go

```diff
@@ -0,0 +1,740 @@
+)
+
+var (
+	ErrTooLarge                 error = fmt.Errorf("Data too large")
+	ErrBufferTooSmall                 = fmt.Errorf("Buffer too small")
+	ErrInvalidData                    = fmt.Errorf("The data does not have the expected format")
+	ErrNoServerEncryption             = fmt.Errorf("The server does not support encryption.")
+	ErrServerRequiresEncryption       = fmt.Errorf("The server requires encryption.")
+	ErrInvalidState                   = fmt.Errorf("Operation cannot be performed in this state")
+)
+
+// https://msdn.microsoft.com/en-us/library/dd358342.aspx
+const (
+	TDSStatusNormal                  uint8 = 0x00
+	TDSStatusEOM                           = 0x01
+	TDSStatusIgnore                        = 0x02
+	TDSStatusResetConnection               = 0x08
+	TDSStatusResetConnectionSkipTran       = 0x10
+)
+
+type TDSPacketType uint8
+
+// https://msdn.microsoft.com/en-us/library/dd304214.aspx
+const (
+	TDSPacketTypeSQLBatch                  TDSPacketType = 0x01
+	TDSPacketTypePreTDS7Login                            = 0x02
+	TDSPacketTypeRPC                                     = 0x03
+	TDSPacketTypeTabularResult                           = 0x04
+	TDSPacketTypeAttentionSignal                         = 0x06
+	TDSPacketTypeBulkLoadData                            = 0x07
+	TDSPacketTypeFederatedAuthToken                      = 0x08
+	TDSPacketTypeTransactionManagerRequest               = 0x0E
+	TDSPacketTypeTDS7Login                               = 0x10
+	TDSPacketTypeSSPI                                    = 0x11
+	TDSPacketTypePrelogin                                = 0x12
+)
+
+// From https://msdn.microsoft.com/en-us/library/dd357559.aspx - PL_OPTION_TOKEN values
+type PreloginOptionToken uint8
+
+const (
+	PreloginVersion         PreloginOptionToken = 0x00
+	PreloginEncryption                          = 0x01
+	PreloginInstance                            = 0x02
+	PreloginThreadID                            = 0x03
+	PreloginMARS                                = 0x04
+	PreloginTraceID                             = 0x05
+	PreloginFedAuthRequired                     = 0x06
+	PreloginNonce                               = 0x07
+	PreloginTerminator                          = 0xFF
+)
+
+// Mapping to documented names, also serves to identify unknown values for JSON marshalling
+var knownPreloginOptionTokens map[PreloginOptionToken]string = map[PreloginOptionToken]string{
+	PreloginVersion:         "VERSION",
+	PreloginEncryption:      "ENCRYPTION",
+	PreloginInstance:        "INSTOPT",
+	PreloginThreadID:        "THREADID",
+	PreloginMARS:            "MARS",
+	PreloginTraceID:         "TRACEID",
+	PreloginFedAuthRequired: "FEDAUTHREQUIRED",
+	PreloginNonce:           "NONCE",
+}
+
+// PreloginOption values are stored as byte arrays; actual types are specified in the docs
+type PreloginOption []byte
+
+// PreloginOptions maps the token to the value for that option
+type PreloginOptions map[PreloginOptionToken]PreloginOption
```

Comment:
**Change**: Since this is just a map, and maps are effectively already pointers (like interfaces), all the functions that operate on `(self *PreloginOptions)` don't need to be pointer receivers, they can just be value receivers. I think this should simply all of those methods